datafusion/optimizer/src/optimizer.rs

```diff
@@ -291,9 +291,13 @@ impl Optimizer {
                 let result = self.optimize_recursively(rule, &new_plan, config);
 
                 match result {
-                    Ok(Some(plan)) => {
-                        assert_schema_is_the_same(rule.name(), &new_plan, &plan)?;
-                        new_plan = plan;
+                    Ok(Some(optimized_plan)) => {
+                        assert_schema_is_the_same(
```

Comment:
I think this assertion should be chained with result first such that the error handling below can kick in and rules that change the schema can also be skipped if needed.